internal/database/database.go

```diff
@@ -0,0 +1,150 @@
+package database
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/atburke/teleport_interview/internal/crypto"
+	"github.com/atburke/teleport_interview/internal/types"
+	_ "github.com/go-sql-driver/mysql"
+	"time"
+)
+
+// tokens and password hash could be []byte objects, but they're strings in the
+// db and in the api, so leaving them as hex strings will save time
+
+// Database represents a database connection.
+type Database interface {
+
+	// CreatePreAuthSession creates a new session for a user that has not yet
+	// authenticated.
+	CreatePreAuthSession(csrfToken string, initTime time.Time) (*types.Session, error)
+
+	// CreateSession creates a new session for an authenticated user.
+	CreateSession(accountId, csrfToken string, initTime time.Time) (*types.Session, error)
+
+	// FetchSession fetches the session identified by the provided token.
+	FetchSession(sessionToken string) (*types.Session, error)
+
+	// DeleteSession deletes a session for a user.
+	DeleteSession(sessionToken string) error
+
+	// FetchAccount fetches the account for a user.
+	FetchAccount(email string) (*types.Account, error)
+
+	// DeleteExpiredSessions deletes sessions that have expired (after either their
+	// idle or absolute timeout has passed)
+	DeleteExpiredSessions(t time.Time) error
+
+	// Close closes any underlying resources.
+	Close()
+}
+
+type MySqlDatabase struct {
+	driver *sql.DB
+}
+
+// let's not bother with reusable prepared statements; the performance impact
+// should be minimal for us
+
+func NewMySqlDatabase(username, password, databaseName string) (*MySqlDatabase, error) {
+	// hardcode non-sensitive info to speed us along
+	driver, err := sql.Open("mysql",
+		fmt.Sprintf("%s:%s@tcp(db:3306)/%s?parseTime=true", username, password, databaseName),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Could not connect to database: %w", err)
+	}
+	return &MySqlDatabase{driver}, nil
+}
+
+func (db *MySqlDatabase) CreatePreAuthSession(csrfToken string, initTime time.Time) (*types.Session, error) {
```

Comment:
Instead of passing in `time.Now()` to each function,. the common Go pattern is to create a fake clock, something like the following:

```go
import (
   "github.com/jonboulle/clockwork"
)

type MySqlDatabase struct {
   driver *sql.DB
   clock clockwork.Clock
}

func NewMySqlDatabase(username, password, databaseName string, clock clockwork.Clock) (*MySqlDatabase, error) {
   [...]
   if clock == nil {
     clock = clockwork.NewRealClock()
   }
   return &MySqlDatabase{
       [...]
       clock: clock,
   }
}
```

Now the caller does not have to worry about time and you can also fake time in tests.